src/libcrun/container.c

```diff
@@ -3706,6 +3770,143 @@ libcrun_container_update_from_values (libcrun_context_t *context, const char *id
+        {
+          // Non-numeric or non-zero value encountered, break the loop
+          break;
+        }
+
+      // Convert capability name to uppercase
+      for (int j = 0; v[j] != '\0'; j++)
+        {
+          v[j] = toupper (v[j]);
+        }
+
+      (*capabilities)[index] = v;
+      index++;
+    }
+
+  (*capabilities)[index] = NULL; // Terminate the array with NULL
+  populate_array_field (&(info->linux.capabilities), *capabilities, *num_capabilities);
+}
+
+void
+retrieve_mount_options (struct features_info_s **info)
+{
+  // Retrieve mount options from wordlist
+  const struct propagation_flags_s *mount_options_list = get_mount_flags_from_wordlist ();
+
+  // Calculate the number of mount options
+  size_t num_mount_options = 0;
+  while (mount_options_list[num_mount_options].name != NULL)
+    {
+      num_mount_options++;
+    }
+
+  // Allocate memory for mount options in info struct
+  (*info)->mount_options = xmalloc0 (num_mount_options * sizeof (char *));
+
+  // Copy mount options to info struct
+  for (size_t i = 0; i < num_mount_options; i++)
+    {
+      (*info)->mount_options[i] = xstrdup (mount_options_list[i].name);
+    }
+}
+
+int
+libcrun_container_get_features (struct features_info_s **info)
+{
+  // Allocate memory for the features_info_s structure
+  *info = xmalloc0 (sizeof (struct features_info_s));
+  // Hardcoded feature information
+  (*info)->oci_version_min = xstrdup ("1.0.0");
+  (*info)->oci_version_max = xstrdup ("1.0.2-dev");
+
+  // Populate hooks
+  size_t num_hooks = sizeof (hooks) / sizeof (hooks[0]);
+  populate_array_field (&((*info)->hooks), hooks, num_hooks);
+
+  // Populate mount_options
+  retrieve_mount_options (info);
+
+  // Populate namespaces
+  size_t num_namspaces = sizeof (namespaces) / sizeof (namespaces[0]);
+  populate_array_field (&((*info)->linux.namespaces), namespaces, num_namspaces);
+
+  // Populate capabilities
+  size_t num_capabilities = 0;
+  const char **capabilities = NULL;
+  populate_capabilities (*info, &capabilities, &num_capabilities);
+
+  // Hardcode the values for cgroup
+  (*info)->linux.cgroup.v1 = true;
+  (*info)->linux.cgroup.v2 = true;
+#ifdef HAVE_SYSTEMD
+  (*info)->linux.cgroup.systemd = true;
+  (*info)->linux.cgroup.systemdUser = true;
+#endif
+
+  // Put seccomp values
+  (*info)->linux.seccomp.enabled = true;
+  // Populate actions
+  size_t num_actions = sizeof (actions) / sizeof (actions[0]);
+  populate_array_field (&((*info)->linux.seccomp.actions), actions, num_actions);
+  // Populate operators
+  size_t num_operators = sizeof (operators) / sizeof (operators[0]);
+  populate_array_field (&((*info)->linux.seccomp.operators), operators, num_operators);
+  // Populate archs
+  size_t num_archs = sizeof (archs) / sizeof (archs[0]);
+  populate_array_field (&((*info)->linux.seccomp.archs), archs, num_archs);
+
+  // Put values for apparmor and selinux
+  (*info)->linux.apparmor.enabled = true;
+  (*info)->linux.selinux.enabled = true;
+
```

Comment:
I'd expect annotation to be added here too:
```json
  "annotations": {
    "io.github.seccomp.libseccomp.version": "2.5.4",
    "org.opencontainers.runc.checkpoint.enabled": "true",
    "io.github.containers.crun.commit": "...",
    "io.github.containers.crun.version": "1.9.0"
  }
```
https://github.com/opencontainers/runtime-spec/blob/v1.1.0-rc.3/features.md#annotations

Reply:
I thought we need to add these annotations:
```sh
$ git grep "## \`run.oci."
crun.1.md:## `run.oci.mount_context_type=context`
crun.1.md:## `run.oci.seccomp.receiver=PATH`
crun.1.md:## `run.oci.seccomp.plugins=PATH`
crun.1.md:## `run.oci.seccomp_fail_unknown_syscall=1`
crun.1.md:## `run.oci.seccomp_bpf_data=PATH`
crun.1.md:## `run.oci.keep_original_groups=1`
crun.1.md:## `run.oci.pidfd_receiver=PATH`
crun.1.md:## `run.oci.systemd.force_cgroup_v1=/PATH`
crun.1.md:## `run.oci.systemd.subgroup=SUBGROUP`
crun.1.md:## `run.oci.delegate-cgroup=DELEGATED-CGROUP`
crun.1.md:## `run.oci.hooks.stdout=FILE`
crun.1.md:## `run.oci.hooks.stderr=FILE`
crun.1.md:## `run.oci.handler=HANDLER`
```

Reply:
"annotations" in the feature JSON is not the list of those annotations

Reply:
- https://github.com/opencontainers/runtime-spec/pull/1212

Reply:
> I'd expect annotation to be added here too:
> 
> ```json
>   "annotations": {
>     "io.github.seccomp.libseccomp.version": "2.5.4",
>     "org.opencontainers.runc.checkpoint.enabled": "true",
>     "io.github.containers.crun.commit": "...",
>     "io.github.containers.crun.version": "1.9.0"
>   }
> ```
> 
> https://github.com/opencontainers/runtime-spec/blob/v1.1.0-rc.3/features.md#annotations

do we need to change--> `org.opencontainers.runc.checkpoint.enabled` to `org.containers.crun.checkpoint.enabled`?